adlfs/transfer.py

```diff
@@ -0,0 +1,312 @@
+        self._transfer = transfer
+        self._merge = merge
+        self._nthreads = nthreads or multiprocessing.cpu_count()
+        self._chunksize = chunksize
+        self._chunkretries = 5
+        self._tmp_path = tmp_path
+        self._tmp_prefix = tmp_prefix
+        self._tmp_unique = tmp_unique
+        self._persist_path = persist_path
+        self._pool = ThreadPoolExecutor(self._nthreads)
+        self._shutdown_event = threading.Event()
+        self._files = {}
+        self._fstates = DisjointState(
+            'pending', 'transferring', 'merging', 'finished', 'cancelled',
+            'errored')
+
+    def submit(self, src, dst, nbytes):
+        """
+        All submitted files start in the `pending` state until `run()` is
+        called.
+        """
+        self._fstates[(src, dst)] = 'pending'
+        self._files[(src, dst)] = dict(
+            nbytes=nbytes,
+            start=None,
+            stop=None,
+            chunks={},
+            cstates=DisjointState('running', 'finished', 'cancelled', 'errored'),
+            merge=None)
+
+    def _submit(self, fn, *args, **kwargs):
+        kwargs['shutdown_event'] = self._shutdown_event
+        return self._pool.submit(fn, *args, **kwargs)
+
+    def _scatter(self, src, dst, transfer):
+        """ Split a given file into chunks """
+        dic = self._files[(src, dst)]
+        self._fstates[(src, dst)] = 'transferring'
+        offsets = list(range(0, dic['nbytes'], self._chunksize))
+        for offset in offsets:
+            if self._tmp_path and len(offsets) > 1:
+                subdir = uuid.uuid1().hex[:10] if self._tmp_unique else ''
+                prefix = self._tmp_prefix or ''
+                name = os.path.join(
+                    self._tmp_path,
+                    subdir,
+                    prefix + str(offset))
+            else:
+                name = dst
+            logger.debug("Submitted %s, byte offset %d", name, offset)
+            dic['cstates'][name] = 'running'
+            dic['chunks'][name] = dict(
+                future=self._submit(transfer, self._adlfs, src, name, offset,
+                                    self._chunksize),
+                retries=self._chunkretries,
+                offset=offset)
+
+    @property
+    def progress(self):
+        files = []
+        for key in self._files:
+            src, dst = key
+            chunks = []
+            for name in self._files[key]['chunks']:
+                chunks.append(Chunk(
+                    name=name,
+                    state=self._files[key]['cstates'][name],
+                    offset=self._files[key]['chunks'][name]['offset'],
+                    retries=self._files[key]['chunks'][name]['retries']))
+            files.append(File(
+                src=src,
+                dst=dst,
+                state=self._fstates[key],
+                nbytes=self._files[key]['nbytes'],
+                start=self._files[key]['start'],
```

Comment:
This will be the start time of the whole process, no? Should maybe be `time.time()`, since there will always be some lag even if it's one of the chunks read to run immediately. Or perhaps update this when the chunk actually starts its running state. (not very important, just a thought)